src/rust/lib/graph-editor/src/component/operator.rs

```diff
@@ -0,0 +1,271 @@
+}
+
+/// A component that owns a Network. Can be used to create external animations.
+pub trait Networked {
+    /// Return a reference to the components network.
+    fn network(&self) -> &frp::Network;
+}
+
+// ==================================
+// === Fullscreen Operator Handle ===
+// ==================================
+
+/// FullscreenOperatorCellHandle is a helper type that wraps a `FullscreenOperator` and applies an
+/// undos the operator on the inner component. This can be used to ensure that only a single
+/// component is made fullscreen at any time.
+#[derive(Debug,CloneRef,Derivative)]
+#[derivative(Clone(bound=""))]
+#[derivative(Default(bound=""))]
+pub struct FullscreenOperatorHandle<T> {
+    operator: Rc<RefCell<Option<FullscreenOperator<T>>>>
+}
+
+impl<T:display::Object+Resizable+ NativeComponent +CloneRef+Networked+'static> FullscreenOperatorHandle<T> {
+    /// returns whether there is a component that is in fullscreen mode.
+    pub fn is_active(&self) -> bool {
+        self.operator.borrow().is_some()
+    }
+
+    /// Enables fullscreen mode for the given component. If there is another component already in
+    /// fullscreen mode, it disables fullscreen for that component.
+    pub fn set_fullscreen(&self, target:T, scene:Scene) {
+        self.disable_fullscreen();
+        let operator = FullscreenOperator::apply(target,scene);
+        self.operator.set(operator);
+    }
+
+    /// Disables fullscreen mode for the given component.
+    pub fn disable_fullscreen(&self) {
+        if let Some(old) = self.operator.borrow_mut().take() {
+            old.undo();
+        }
+    }
+
+    /// Return a ref clone of the fullscreen element.
+    pub fn get_element(&self) -> Option<T>{
+        self.operator.borrow().as_ref().map(|op| op.target.clone_ref())
+    }
+}
+
+
+
+// ===============================
+// === Layer Management Helper ===
+// ===============================
+
+/// Indicates the required target layer.
+// FIXME this is a layer management hack. Remove this once we have nicer scene layer management.
+#[derive(Debug)]
+pub enum SymbolType {
+    /// A symbol that goes onto the `Main` layer.
+    Main (Symbol),
+    /// A visualisation symbol that goes above the `Main` layer, but below the cursor.
+    Visualisation (Symbol),
+}
+
+/// Moves the given components shapes to the default scene layers.
+// FIXME This is an ugly hack for layer management.
+// FIXME Needs to be removed as soon as we have something better.
+pub fn set_layers_normal<T: NativeComponent>(target:&T, scene:&Scene){
+    target.unset_layers_all(&scene);
+    for symbol in target.symbols() {
+        match symbol {
+            SymbolType::Main(symbol)          => scene.views.main.add(&symbol),
+            SymbolType::Visualisation(symbol) => scene.views.visualisation.add(&symbol),
+        }
+    }
+}
+
+/// Moves the given components shapes to the fullscreen scene layers.
+// FIXME This is an ugly hack for layer management.
```

Comment:
Describe the hack please. What is the hack? Or, why you consider it to be a hack?